# レポート課題(その1)

を解くためのノート

In [2]:
def gcd(a, b):
    if(b == 0):
        return a
    return gcd(b,a%b)

gcd(111, 30)

3

In [3]:
def mod(a0, a1):
    return (a0%a1)

# 拡張ユークリッドの互除法
# input: 二つの整数a, b
# output: d=gcd(a, b)とax+by = dを満たす整数x, y
def ex_euclid(a, b):
    if abs(a) < abs(b): # a, bの絶対値を比較(abs(a) = |a|, abs(b) = |b|)
        a, b = b, a     # |a| < |b| の場合は, a(割られる数)とb(割る数)を入れ替える
        change_flag = 1 # a, bの入れ替えがある場合とない場合とで、出力が変わるので、入れ替えがあったかどうかを記録
    else:
        change_flag = 0 # a, bの入れ替えがある場合とない場合とで、出力が変わるので、入れ替えがあったかどうかを記録

    a0, a1 = a, b
    #x0 -> prevx, y0 -> prevy
    x0, x1 = 1, 0
    y0, y1 = 0, 1
    while True:
        if a1 == 0:
            if change_flag == 0:
                return [a0, x0, y0] # a, bの入れ替えがない場合はそのまま出力
            else:
                return [a0, y0, x0] # a, bを入れ替えた場合はy0, x0を入れ替えて出力

        # 以下の "=" の後にコードを書いて関数を完成させてください。
        # qはa0をa1で割った時の商で, rはその時の剰余です。
        r = mod(a0, a1)
        q = (a0-r)//a1
        x0, x1 = x1, x0 - q * x1
        y0, y1 = y1, y0 - q * y1
        a0 = a1
        a1 = r
        
ex_euclid(111, 30)


[3, 3, -11]

## 参考にした資料

[Python3.0による拡張ユークリッドの互除法](https://teratail.com/questions/263807)

[拡張ユークリッドの互除法 〜 一次不定方程式 ax + by = c の解き方 〜](https://qiita.com/drken/items/b97ff231e43bce50199a)

[「拡張ユークリッドの互除法」の実装](https://misteer.hatenablog.com/entry/extgcd)

[拡張ユークリッドの互除法](https://algo-logic.info/extend-gcd/)

[Extended Euclidean Algorithm](http://anh.cs.luc.edu/331/notes/xgcd.pdf)

In [4]:
X = 27182818
Y = 10000000
print(X+Y)

37182818


In [5]:
def gcd_GenTable(x ,y):
    if( y == 0):
        return x
    print(f"|${x}/{y} = {x//y} あまり {x%y}$|$ {x} = {x//y} \\times {y} + {x%y}$|$ {x%y} = {x} - {x//y} \\times {y}$|")
    return(gcd_GenTable(y, x%y))
print("|割り算|かけ算の式に直す|さらに式変形する|")
print("|----|----|----|")
gcd_GenTable(111, 30)

|割り算|かけ算の式に直す|さらに式変形する|
|----|----|----|
|$111/30 = 3 あまり 21$|$ 111 = 3 \times 30 + 21$|$ 21 = 111 - 3 \times 30$|
|$30/21 = 1 あまり 9$|$ 30 = 1 \times 21 + 9$|$ 9 = 30 - 1 \times 21$|
|$21/9 = 2 あまり 3$|$ 21 = 2 \times 9 + 3$|$ 3 = 21 - 2 \times 9$|
|$9/3 = 3 あまり 0$|$ 9 = 3 \times 3 + 0$|$ 0 = 9 - 3 \times 3$|


3

|割り算|かけ算の式に直す|さらに式変形する|
|----|----|----|
|$111/30 = 3 あまり 21$|$ 111 = 3 \times 30 + 21$|$ 21 = 111 - 3 \times 30$|
|$30/21 = 1 あまり 9$|$ 30 = 1 \times 21 + 9$|$ 9 = 30 - 1 \times 21$|
|$21/9 = 2 あまり 3$|$ 21 = 2 \times 9 + 3$|$ 3 = 21 - 2 \times 9$|
|$9/3 = 3 あまり 0$|$ 9 = 3 \times 3 + 0$|$ 0 = 9 - 3 \times 3$|

In [6]:
print("|割り算|かけ算の式に直す|さらに式変形する|")
print("|----|----|----|")
gcd_GenTable(X, Y)

|割り算|かけ算の式に直す|さらに式変形する|
|----|----|----|
|$27182818/10000000 = 2 あまり 7182818$|$ 27182818 = 2 \times 10000000 + 7182818$|$ 7182818 = 27182818 - 2 \times 10000000$|
|$10000000/7182818 = 1 あまり 2817182$|$ 10000000 = 1 \times 7182818 + 2817182$|$ 2817182 = 10000000 - 1 \times 7182818$|
|$7182818/2817182 = 2 あまり 1548454$|$ 7182818 = 2 \times 2817182 + 1548454$|$ 1548454 = 7182818 - 2 \times 2817182$|
|$2817182/1548454 = 1 あまり 1268728$|$ 2817182 = 1 \times 1548454 + 1268728$|$ 1268728 = 2817182 - 1 \times 1548454$|
|$1548454/1268728 = 1 あまり 279726$|$ 1548454 = 1 \times 1268728 + 279726$|$ 279726 = 1548454 - 1 \times 1268728$|
|$1268728/279726 = 4 あまり 149824$|$ 1268728 = 4 \times 279726 + 149824$|$ 149824 = 1268728 - 4 \times 279726$|
|$279726/149824 = 1 あまり 129902$|$ 279726 = 1 \times 149824 + 129902$|$ 129902 = 279726 - 1 \times 149824$|
|$149824/129902 = 1 あまり 19922$|$ 149824 = 1 \times 129902 + 19922$|$ 19922 = 149824 - 1 \times 129902$|
|$129902/19922 = 6 あまり 10370$|$ 129902 = 6 \times 19

2

|割り算|かけ算の式に直す|さらに式変形する|
|----|----|----|
|$27182818/10000000 = 2 あまり 7182818$|$ 27182818 = 2 \times 10000000 + 7182818$|$ 7182818 = 27182818 - 2 \times 10000000$|
|$10000000/7182818 = 1 あまり 2817182$|$ 10000000 = 1 \times 7182818 + 2817182$|$ 2817182 = 10000000 - 1 \times 7182818$|
|$7182818/2817182 = 2 あまり 1548454$|$ 7182818 = 2 \times 2817182 + 1548454$|$ 1548454 = 7182818 - 2 \times 2817182$|
|$2817182/1548454 = 1 あまり 1268728$|$ 2817182 = 1 \times 1548454 + 1268728$|$ 1268728 = 2817182 - 1 \times 1548454$|
|$1548454/1268728 = 1 あまり 279726$|$ 1548454 = 1 \times 1268728 + 279726$|$ 279726 = 1548454 - 1 \times 1268728$|
|$1268728/279726 = 4 あまり 149824$|$ 1268728 = 4 \times 279726 + 149824$|$ 149824 = 1268728 - 4 \times 279726$|
|$279726/149824 = 1 あまり 129902$|$ 279726 = 1 \times 149824 + 129902$|$ 129902 = 279726 - 1 \times 149824$|
|$149824/129902 = 1 あまり 19922$|$ 149824 = 1 \times 129902 + 19922$|$ 19922 = 149824 - 1 \times 129902$|
|$129902/19922 = 6 あまり 10370$|$ 129902 = 6 \times 19922 + 10370$|$ 10370 = 129902 - 6 \times 19922$|
|$19922/10370 = 1 あまり 9552$|$ 19922 = 1 \times 10370 + 9552$|$ 9552 = 19922 - 1 \times 10370$|
|$10370/9552 = 1 あまり 818$|$ 10370 = 1 \times 9552 + 818$|$ 818 = 10370 - 1 \times 9552$|
|$9552/818 = 11 あまり 554$|$ 9552 = 11 \times 818 + 554$|$ 554 = 9552 - 11 \times 818$|
|$818/554 = 1 あまり 264$|$ 818 = 1 \times 554 + 264$|$ 264 = 818 - 1 \times 554$|
|$554/264 = 2 あまり 26$|$ 554 = 2 \times 264 + 26$|$ 26 = 554 - 2 \times 264$|
|$264/26 = 10 あまり 4$|$ 264 = 10 \times 26 + 4$|$ 4 = 264 - 10 \times 26$|
|$26/4 = 6 あまり 2$|$ 26 = 6 \times 4 + 2$|$ 2 = 26 - 6 \times 4$|
|$4/2 = 2 あまり 0$|$ 4 = 2 \times 2 + 0$|$ 0 = 4 - 2 \times 2$|

In [7]:
ex_euclid(X, Y)

[2, -2310511, 6280620]

In [8]:
import time
import copy

def divByList1(list1, list2):
    return_list = copy.copy(list2)
    return_list[0] = list1[1]//list2[1]
    return_list[4] = list1[3]//list2[3]
    return(return_list)

def divByList2(list1, list2):
    return_list = copy.deepcopy(list2)


    return_list[1] = list1[1]%list2[1]
    return_list[2] = list1[2]%list2[2]
    return_list[3] = list1[3]%list2[3]

    return_list[0] = list2[1]//return_list[1]
    return_list[4] = list2[3]//return_list[3]

    return_list[5][0] = list1[5][0] - list2[4] * list2[5][0]
    return_list[5][1] = list1[5][1] - list2[4] * list2[5][1]
    

    return(return_list)

def printTableHead(list1):
    retStr1 = "|"
    retStr2 = "|"
    for content in list1:
        retStr1 += "xxx|"
        retStr2 += "----|"
    print(retStr1)
    print(retStr2)

def printTableList(list1):
    print(f"|", end="")
    for content in list1:
        print(f"{content}|", end="")
    print("")

def genLehmerTable(list1, list2):
    list_1 = copy.deepcopy(list1)
    list_2 = copy.deepcopy(list2)
    while(list_2[0] == list_2[4] and list_2[2] != 0):
        list_tmp = divByList2(list_1, list_2)
        printTableList(list_tmp)
        list_1 = copy.deepcopy(list_2)
        list_2 = copy.deepcopy(list_tmp)
        # time.sleep(1)

In [9]:
X = 906987930030
Y = 860577930030

# X = 27182818
# Y = 10000000

list1 = [0, int(str(X)[:4]),    X,  int(str(X)[:4])+1,  0, [1,0]]
list2 = [0,   int(str(Y)[:4])+1,  Y,  int(str(Y)[:4]),    0, [0,1]]
list2 = divByList1(list1, list2)

printTableHead(list1)
printTableList(list1)
printTableList(list2)
genLehmerTable(list1, list2)



|xxx|xxx|xxx|xxx|xxx|xxx|
|----|----|----|----|----|----|
|0|9069|906987930030|9070|0|[1, 0]|
|1|8606|860577930030|8605|1|[0, 1]|
|18|463|46410000000|465|18|[1, -1]|
|1|272|25197930030|235|1|[-18, 19]|
|1|191|21212069970|230|1|[19, -20]|
|2|81|3985860060|5|46|[-37, 39]|


In [10]:
list1 = [1,272,25197930030,235,1,[-18, 19]]
list2 = [1,191,21212069970,230,1,[19, -20]]
list1[1] = int(str(list1[2])[:4])
list1[3] = int(str(list1[2])[:4])+1
list2[1] = int(str(list2[2])[:4])
list2[3] = int(str(list2[2])[:4])+1
printTableList(list1)
printTableList(list2)
genLehmerTable(list1, list2)



|1|2519|25197930030|2520|1|[-18, 19]|
|1|2121|21212069970|2122|1|[19, -20]|
|5|398|3985860060|398|5|[-37, 39]|
|3|131|1282769670|132|3|[204, -215]|
|26|5|137551050|2|66|[-649, 684]|


In [11]:
list1 = [5,398,3985860060,398,5,[-37, 39]]
list2 = [3,131,1282769670,132,3,[204, -215]]
list1[1] = int(str(list1[2])[:4])
list1[3] = int(str(list1[2])[:4])+1
list2[1] = int(str(list2[2])[:4])
list2[3] = int(str(list2[2])[:4])+1
printTableList(list1)
printTableList(list2)
genLehmerTable(list1, list2)



|5|3985|3985860060|3986|5|[-37, 39]|
|3|1282|1282769670|1283|3|[204, -215]|
|9|139|137551050|137|9|[-649, 684]|
|4|31|44810220|50|2|[6045, -6371]|


In [12]:
list1 = [3,1282,1282769670,1283,3,[204, -215]]
list2 = [9,139,137551050,137,9,[-649, 684]]
list1[1] = int(str(list1[2])[:4])
list1[3] = int(str(list1[2])[:4])+1
list2[1] = int(str(list2[2])[:4])
list2[3] = int(str(list2[2])[:4])+1
printTableList(list1)
printTableList(list2)
genLehmerTable(list1, list2)



|3|1282|1282769670|1283|3|[204, -215]|
|9|1375|137551050|1376|9|[-649, 684]|
|1|1282|44810220|1283|1|[6045, -6371]|
|13|93|3120390|93|13|[-6694, 7055]|
|1|73|1124760|74|1|[93067, -98086]|
|3|20|870870|19|3|[-99761, 105141]|
|1|13|253890|17|1|[392350, -413509]|
|1|7|109200|2|8|[-492111, 518650]|


In [13]:
list1 = [3,20,870870,19,3,[-99761, 105141]]
list2 = [1,13,253890,17,1,[392350, -413509]]
list1[1] = int(str(list1[2])[:4])
list1[3] = int(str(list1[2])[:4])+1
list2[1] = int(str(list2[2])[:4])
list2[3] = int(str(list2[2])[:4])+1
printTableList(list1)
printTableList(list2)
genLehmerTable(list1, list2)


|3|8708|870870|8709|3|[-99761, 105141]|
|1|2538|253890|2539|1|[392350, -413509]|
|2|1094|109200|1092|2|[-492111, 518650]|
|3|350|35490|355|3|[1376572, -1450809]|
|7|44|2730|27|13|[-4621827, 4871077]|


In [14]:

list1 = [2,1094,109200,1092,2,[-492111, 518650]]
list2 = [3,350,35490,355,3,[1376572, -1450809]]
list1[1] = int(str(list1[2])[:4])
list1[3] = int(str(list1[2])[:4])+1
list2[1] = int(str(list2[2])[:4])
list2[3] = int(str(list2[2])[:4])+1
printTableList(list1)
printTableList(list2)
genLehmerTable(list1, list2)

|2|1092|109200|1093|2|[-492111, 518650]|
|3|3549|35490|3550|3|[1376572, -1450809]|
|3|1092|2730|1093|3|[-4621827, 4871077]|
|4|273|0|271|4|[15242053, -16064040]|


__ 上のセルでの出力をMarkDownとして出力したもの __


|xxx|xxx|xxx|xxx|xxx|xxx|
|----|----|----|----|----|----|
|0|9069|906987930030|9070|0|[1, 0]|
|1|8606|860577930030|8605|1|[0, 1]|
|18|463|46410000000|465|18|[1, -1]|
|1|272|25197930030|235|1|[-18, 19]|
|1|191|21212069970|230|1|[19, -20]|
|2|81|3985860060|5|46|[-37, 39]|
|1|2519|25197930030|2520|1|[-18, 19]|
|1|2121|21212069970|2122|1|[19, -20]|
|5|398|3985860060|398|5|[-37, 39]|
|3|131|1282769670|132|3|[204, -215]|
|26|5|137551050|2|66|[-649, 684]|
|5|3985|3985860060|3986|5|[-37, 39]|
|3|1282|1282769670|1283|3|[204, -215]|
|9|139|137551050|137|9|[-649, 684]|
|4|31|44810220|50|2|[6045, -6371]|
|3|1282|1282769670|1283|3|[204, -215]|
|9|1375|137551050|1376|9|[-649, 684]|
|1|1282|44810220|1283|1|[6045, -6371]|
|13|93|3120390|93|13|[-6694, 7055]|
|1|73|1124760|74|1|[93067, -98086]|
|3|20|870870|19|3|[-99761, 105141]|
|1|13|253890|17|1|[392350, -413509]|
|1|7|109200|2|8|[-492111, 518650]|
|3|8708|870870|8709|3|[-99761, 105141]|
|1|2538|253890|2539|1|[392350, -413509]|
|2|1094|109200|1092|2|[-492111, 518650]|
|3|350|35490|355|3|[1376572, -1450809]|
|7|44|2730|27|13|[-4621827, 4871077]|
|2|1092|109200|1093|2|[-492111, 518650]|
|3|3549|35490|3550|3|[1376572, -1450809]|
|3|1092|2730|1093|3|[-4621827, 4871077]|
|4|273|0|271|4|[15242053, -16064040]|

In [15]:
def BinaryAlgorithm(num1, num2):
    num_1 = num1
    num_2 = num2
    print(f"初期値 $B_1 = {int(num_1)}, B_2 = {int(num_2)}$\n")
    k = 0
    print(f"$B_1$ , $B_2$ のどちらかが奇数になるまで両方を$2$で割る")
    while(num_1%2 == 0 and num_2%2 == 0):
        k += 1
        num_1 /= 2
        num_2 /= 2
    w = 1
    print(f"その結果、 $B_1 = {int(num_1)}, B_2 = {int(num_2)}$ となった\n")
    while(w != 0):
        w = num_1 - num_2
        print(f"$w = B_1 - B_2$ より $w = {int(w)}$\n")
        if(w == 0):
            print(f"$w == {int(w)} $なのでループは終了。\n")
            break
        print(f"$w$が奇数になるまで$2$で割る。\n")
        while(w % 2 == 0):
            w /= 2
        print(f"$w = 2^s t$ とすると $t = {int(w)}$ となった。\n")
        if(w > 0):
            print(f"$t > 0$ なので $B_1$ に ${int(w)}$ を代入する。\n")
            num_1 = w
        else:
            print(f"$t <= 0$ なので $B_2$ に ${int(w)}$ を代入する。\n")
            num_2 = -w
        time.sleep(2)
    return(2**k * num_2)

In [16]:
X = 910
Y = 555
print(BinaryAlgorithm(X, Y))

初期値 $B_1 = 910, B_2 = 555$

$B_1$ , $B_2$ のどちらかが奇数になるまで両方を$2$で割る
その結果、 $B_1 = 910, B_2 = 555$ となった

$w = B_1 - B_2$ より $w = 355$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = 355$ となった。

$t > 0$ なので $B_1$ に $355$ を代入する。

$w = B_1 - B_2$ より $w = -200$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = -25$ となった。

$t <= 0$ なので $B_2$ に $-25$ を代入する。

$w = B_1 - B_2$ より $w = 330$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = 165$ となった。

$t > 0$ なので $B_1$ に $165$ を代入する。

$w = B_1 - B_2$ より $w = 140$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = 35$ となった。

$t > 0$ なので $B_1$ に $35$ を代入する。

$w = B_1 - B_2$ より $w = 10$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = 5$ となった。

$t > 0$ なので $B_1$ に $5$ を代入する。

$w = B_1 - B_2$ より $w = -20$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = -5$ となった。

$t <= 0$ なので $B_2$ に $-5$ を代入する。

$w = B_1 - B_2$ より $w = 0$

$w == 0 $なのでループは終了。

5.0


In [17]:
X = 915
Y = 555
print(f"以上より、X = {X}, Y = {Y} のとき、GCD(X, Y) = {int(BinaryAlgorithm(X, Y))} となった。")

初期値 $B_1 = 915, B_2 = 555$

$B_1$ , $B_2$ のどちらかが奇数になるまで両方を$2$で割る
その結果、 $B_1 = 915, B_2 = 555$ となった

$w = B_1 - B_2$ より $w = 360$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = 45$ となった。

$t > 0$ なので $B_1$ に $45$ を代入する。

$w = B_1 - B_2$ より $w = -510$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = -255$ となった。

$t <= 0$ なので $B_2$ に $-255$ を代入する。

$w = B_1 - B_2$ より $w = -210$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = -105$ となった。

$t <= 0$ なので $B_2$ に $-105$ を代入する。

$w = B_1 - B_2$ より $w = -60$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = -15$ となった。

$t <= 0$ なので $B_2$ に $-15$ を代入する。

$w = B_1 - B_2$ より $w = 30$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = 15$ となった。

$t > 0$ なので $B_1$ に $15$ を代入する。

$w = B_1 - B_2$ より $w = 0$

$w == 0 $なのでループは終了。

以上より、X = 915, Y = 555 のとき、GCD(X, Y) = 15 となった。


# MarkDownによるレポートの下書き

<style type="text/css">
table {
    border-collapse: collapse;
}
table, th, td {
    border: 1px solid black;
}
</style>

### [練習問題 2-1]

問題を解くにあたって次のような表を作成した。

|割り算|かけ算の式に直す|さらに式変形する|
|----|----|----|
|$27182818/10000000 = 2 あまり 7182818$|$ 27182818 = 2 \times 10000000 + 7182818$|$ 7182818 = 27182818 - 2 \times 10000000$|
|$10000000/7182818 = 1 あまり 2817182$|$ 10000000 = 1 \times 7182818 + 2817182$|$ 2817182 = 10000000 - 1 \times 7182818$|
|$7182818/2817182 = 2 あまり 1548454$|$ 7182818 = 2 \times 2817182 + 1548454$|$ 1548454 = 7182818 - 2 \times 2817182$|
|$2817182/1548454 = 1 あまり 1268728$|$ 2817182 = 1 \times 1548454 + 1268728$|$ 1268728 = 2817182 - 1 \times 1548454$|
|$1548454/1268728 = 1 あまり 279726$|$ 1548454 = 1 \times 1268728 + 279726$|$ 279726 = 1548454 - 1 \times 1268728$|
|$1268728/279726 = 4 あまり 149824$|$ 1268728 = 4 \times 279726 + 149824$|$ 149824 = 1268728 - 4 \times 279726$|
|$279726/149824 = 1 あまり 129902$|$ 279726 = 1 \times 149824 + 129902$|$ 129902 = 279726 - 1 \times 149824$|
|$149824/129902 = 1 あまり 19922$|$ 149824 = 1 \times 129902 + 19922$|$ 19922 = 149824 - 1 \times 129902$|
|$129902/19922 = 6 あまり 10370$|$ 129902 = 6 \times 19922 + 10370$|$ 10370 = 129902 - 6 \times 19922$|
|$19922/10370 = 1 あまり 9552$|$ 19922 = 1 \times 10370 + 9552$|$ 9552 = 19922 - 1 \times 10370$|
|$10370/9552 = 1 あまり 818$|$ 10370 = 1 \times 9552 + 818$|$ 818 = 10370 - 1 \times 9552$|
|$9552/818 = 11 あまり 554$|$ 9552 = 11 \times 818 + 554$|$ 554 = 9552 - 11 \times 818$|
|$818/554 = 1 あまり 264$|$ 818 = 1 \times 554 + 264$|$ 264 = 818 - 1 \times 554$|
|$554/264 = 2 あまり 26$|$ 554 = 2 \times 264 + 26$|$ 26 = 554 - 2 \times 264$|
|$264/26 = 10 あまり 4$|$ 264 = 10 \times 26 + 4$|$ 4 = 264 - 10 \times 26$|
|$26/4 = 6 あまり 2$|$ 26 = 6 \times 4 + 2$|$ 2 = 26 - 6 \times 4$|
|$4/2 = 2 あまり 0$|$ 4 = 2 \times 2 + 0$|$ 0 = 4 - 2 \times 2$|
 

この表を基に $ 2 = 26 - 6 \times 4 $ を次のように変換を繰り返した。

$$
    2 = 26 - 6 \times 4
$$

$ 4 = 264 - 10 \times 26 $ を代入。

$$ 
    2 = 61 \times 26 - 6 \times 264
$$

$ 26 = 554 - 2 \times 264 $ を代入。

$$ 
    2 = 61 \times 554 - 128 \times 264
$$

$ 264 = 818 - 1 \times 554 $ を代入。

$$
    2 = 189 \times 553 - 128 \times 8181
$$

$ 554 = 9552 - 11 \times 818 $ を代入。

$$
    2 = 189 \times 9552 - 2207 \times 818
$$

$ 818 = 10370 - 1 \times 9552 $ を代入。

$$
    2 = 2396 \times 9552 - 2207 \times 10370
$$

$ 9552 = 19922 - 10370 $ を代入。

$$
    2 = 2396 \times 19922 - 4603 \times 10370
$$

$ 10370 = 129902 - 6 \times 19922 $ を代入。

$$
    2 = 30014 \times 19922 - 4603 \times 129902
$$

$ 19922 = 149824 - 1 \times 129902 $ を代入。

$$
    2 = 30014 \times 149824 - 34617 \times 129902
$$

$ 129902 = 279726 - 1 \times 149824 $ を代入。

$$ 
    2 = 64631 \times 149824 - 34617 \times 279726
$$

$ 149824 = 1268728 - 4 \times 279726 $ を代入。

$$
    2 = 64631 \times 1268728 - 293141 \times 279726
$$

$ 279726 = 154454 - 1 \times 1268728 $ を代入。

$$
    2 = 357772 \times 1268728 - 293141 \times 1548454
$$

$ 1268728 = 2817182 - 1 \times 1548454 $ を代入。

$$
    2 = 357772 \times 2817182 - 650913 \times 1548454
$$

$ 154854 = 7182818 - 2 \times 2817182 $ を代入。

$$
    2 = 1659598 \times 2817182 - 650913 \times 7182818
$$

$ 2817182 = 10000000 - 1 \times 7182818 $ を代入。

$$
    2 = -2310511 \times 7182818 + 1659598 \times 10000000
$$

$ 7182818 = 27182818 -2 \times 10000000 $ を代入。

$$
    2 = 6280620 \times 10000000 - 2310511 \times 27182818
$$

以上より

$ X = 27182818, Y = 10000000 $ のとき、次式を満たすa, bの組み合わせは

$$
    aX + bY = gcd(X,Y)
$$

$ a = -2310511, b = 6280620$ である。

### [練習問題 2-1]

授業資料7ページにあるような表を作成した。次のようになった。また、$\hat{a_i}'$の値と$\hat{a_i}''$の値が異なってしまった場合は、続きの行をその行の二行分上の値で計算しなおすことにしている。

|$q_i'$|$\hat{a_i}'$|$a_i$|$\hat{a_i}''$|$q_i''$|$(\gamma_i, \delta_i)$|
|----|----|----|----|----|----|
|0|9069|906987930030|9070|0|[1, 0]|
|1|8606|860577930030|8605|1|[0, 1]|
|18|463|46410000000|465|18|[1, -1]|
|1|272|25197930030|235|1|[-18, 19]|
|1|191|21212069970|230|1|[19, -20]|
|2|81|3985860060|5|46|[-37, 39]|
|1|2519|25197930030|2520|1|[-18, 19]|
|1|2121|21212069970|2122|1|[19, -20]|
|5|398|3985860060|398|5|[-37, 39]|
|3|131|1282769670|132|3|[204, -215]|
|26|5|137551050|2|66|[-649, 684]|
|5|3985|3985860060|3986|5|[-37, 39]|
|3|1282|1282769670|1283|3|[204, -215]|
|9|139|137551050|137|9|[-649, 684]|
|4|31|44810220|50|2|[6045, -6371]|
|3|1282|1282769670|1283|3|[204, -215]|
|9|1375|137551050|1376|9|[-649, 684]|
|1|1282|44810220|1283|1|[6045, -6371]|
|13|93|3120390|93|13|[-6694, 7055]|
|1|73|1124760|74|1|[93067, -98086]|
|3|20|870870|19|3|[-99761, 105141]|
|1|13|253890|17|1|[392350, -413509]|
|1|7|109200|2|8|[-492111, 518650]|
|3|8708|870870|8709|3|[-99761, 105141]|
|1|2538|253890|2539|1|[392350, -413509]|
|2|1094|109200|1092|2|[-492111, 518650]|
|3|350|35490|355|3|[1376572, -1450809]|
|7|44|2730|27|13|[-4621827, 4871077]|
|2|1092|109200|1093|2|[-492111, 518650]|
|3|3549|35490|3550|3|[1376572, -1450809]|
|3|1092|2730|1093|3|[-4621827, 4871077]|
|4|273|0|271|4|[15242053, -16064040]|

この表より、最大公約数は $ 2730 $ であるとわかった。

### [練習問題 2-3]

授業資料11ページにあるアルゴリズムを $B_1 = 915, B_2 = 555$ として実行した。

初期値 $B_1 = 915, B_2 = 555$

$B_1$ , $B_2$ のどちらかが奇数になるまで両方を$2$で割る
その結果、 $B_1 = 915, B_2 = 555$ となった

$w = B_1 - B_2$ より $w = 360$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = 45$ となった。

$t > 0$ なので $B_1$ に $45$ を代入する。

$w = B_1 - B_2$ より $w = -510$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = -255$ となった。

$t <= 0$ なので $B_2$ に $-255$ を代入する。

$w = B_1 - B_2$ より $w = -210$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = -105$ となった。

$t <= 0$ なので $B_2$ に $-105$ を代入する。

$w = B_1 - B_2$ より $w = -60$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = -15$ となった。

$t <= 0$ なので $B_2$ に $-15$ を代入する。

$w = B_1 - B_2$ より $w = 30$

$w$が奇数になるまで$2$で割る。

$w = 2^s t$ とすると $t = 15$ となった。

$t > 0$ なので $B_1$ に $15$ を代入する。

$w = B_1 - B_2$ より $w = 0$

$w == 0 $なのでループは終了。

以上より、X = 915, Y = 555 のとき、GCD(X, Y) = 15 となった。